lib/folio_client/token_wrapper.rb

```diff
@@ -4,7 +4,7 @@ class FolioClient
   # Wraps API operations to request new access token if expired
   class TokenWrapper
     def self.refresh(config, connection)
-      yield
+      yield.tap { |response| UnexpectedResponse.call(response) unless response.success? }
```

Comment:
i was on the fence about whether to use `UnexpectedResponse.call` to throw the same `UnauthorizedError`, keeping the `rescue` below, vs changing to a conditional on `response.status == 401` and getting rid of the `rescue` for flow control.  happy to do it the other way, if folks prefer that.

Reply:
My naïve question is what happens w/ no changes to `TokenWrapper` in this PR? I'm fuzzy on why this change is needed.

Reply:
without changing `TokenWrapper`, there's no way for it to detect rejection of the authN token, because no exception is raised and as it is, it doesn't examine the response code.  so it'd just return the 401 to the caller.

e.g., if i just get rid of this change but leave the PR otherwise the same, the one relevant test fails:
```
Failures:

  1) FolioClient when token is expired fetches new token and retries
     Failure/Error: raise UnauthorizedError, "There was a problem with the access token: #{response.body}"
     
     FolioClient::UnauthorizedError:
       There was a problem with the access token: invalid authN token
     # ./lib/folio_client/unexpected_response.rb:10:in `call'
     # ./lib/folio_client.rb:65:in `get'
     # ./lib/folio_client/inventory.rb:17:in `fetch_hrid'
     # ./lib/folio_client.rb:94:in `fetch_hrid'
     # ./lib/folio_client.rb:52:in `fetch_hrid'
     # ./spec/folio_client_spec.rb:167:in `block (4 levels) in <top (required)>'
     # ./spec/folio_client_spec.rb:167:in `block (3 levels) in <top (required)>'
```

this was actually what i did at first, causing the relevant test from pre- #26 `main` to fail, because the `UnauthorizedError` would just bubble up to the `UnexpectedResponse` handler after the `TokenWrapper` request block.  feeling like this change shouldn't necessitate a spec change was what motivated #26.  i just recreated the old error by making just this PR's change, minus the `TokenWrapper.refresh` change, on the pre #26 

```
Failures:

  1) FolioClient when token is expired fetches new token and retries
     Failure/Error: (v == :raise) ? raise(FolioClient::UnauthorizedError) : v
     
     FolioClient::UnauthorizedError:
       FolioClient::UnauthorizedError
     # ./spec/folio_client_spec.rb:123:in `block (4 levels) in <top (required)>'
     # ./lib/folio_client.rb:95:in `fetch_hrid'
     # ./lib/folio_client.rb:52:in `fetch_hrid'
     # ./spec/folio_client_spec.rb:128:in `block (4 levels) in <top (required)>'
     # ./spec/folio_client_spec.rb:128:in `block (3 levels) in <top (required)>'
```

(yay for the old tests catching that, though!)

Reply:
before this change, the `UnexpectedResponse` handler in the `get` and `post` helper methods would've raised the error that `TokenWrapper.refresh` rescues

Reply:
or, more succinctly, faraday's GET and POST methods don't raise on non-200 response codes 🙂

Reply:
My head is spinning a bit. Unless someone else groks and approves, I'd probably opt for pairing to walk through it real quick. Also, do the methods like `fetch_hrid` still auto-refresh tokens? I.e., are they using the `get` and `post` methods or are they calling those directly on the `connection`? I think supporting those two modes of operation in the client is part of what's twisting my wig.

Reply:
no worries 😆 

> Unless someone else groks and approves, I'd probably opt for pairing to walk through it real quick.

attending a talk rn, but happy to pair after (should be free around 4:10), or tomorrow morning if it's too late in the day today.

> Also, do the methods like `fetch_hrid` still auto-refresh tokens?

yes behaviorally, but...

> I.e., are they using the `get` and `post` methods or are they calling those directly on the connection?

they don't have to know about token expiry, since the higher-level methods no longer deal directly with auto-refreshing tokens themselves.  they all rely on the centralized http helper methods from `client`.  afaict, the only direct `connection.get` and `connection.post` calls are in these http helper methods, and in the `Authenticator#token` method (i suspect that `Authenticator` is the one place that should use `connection` directly outside of the request helper methods, since the method that explicitly gets a new token shouldn't need to worry about token expiry).

Reply:
```
% git grep connection.post
lib/folio_client.rb:      connection.post(path, request, {"x-okapi-token": config.token})
lib/folio_client/authenticator.rb:      response = connection.post("/authn/login", login_params.to_json)
% git grep connection.get
lib/folio_client.rb:      connection.get(path, params, {"x-okapi-token": config.token})
%
% git grep TokenWrapper.refresh
lib/folio_client.rb:    response = TokenWrapper.refresh(config, connection) do
lib/folio_client.rb:    response = TokenWrapper.refresh(config, connection) do
```

Reply:
OK, I see, thanks for walking through that. 😄

Reply:
np!